(TK-334) debug logging for restart and shutdown

cprice404 opened the pull request:
This commit adds/improves some debugging statements to clarify what
is happening with lifecycle events inside of TK, and when we shut
down the Clojure agent pool.
    
To achieve this we needed to provide a bit of schema around the
items that are put onto the core.async channel for lifecycle management.

fpringvaldsen commented:
:+1:

rlinehan commented on src/puppetlabs/trapperkeeper/internal.clj:
```diff
@@ -212,15 +224,22 @@
+                        (log/debugf "Lifecycle worker completed %s lifecycle task; awaiting next task." type)
+                        (recur))
+
+                      (throw (IllegalStateException. "Unrecognized lifecycle task:" task))))))
```
It doesn't seem like it will be possible to get here, since the schema/validate on line 227 (https://github.com/puppetlabs/trapperkeeper/pull/223/commits/59347aa16f46a61fa77a41d4518e9fe728543f29#diff-452afe9e02bf063094d280de34cf34bcR227) will validate that `:type` is one of `:shutdown`, `:boot`, and `:restart`.

cprice404 replied:
I agree, but I still consider it best practice to have a default case in a switch statement.  Seems harmless?

Maybe someone changes the schema in the future and doesn't update this switch, and the default case becomes reachable?

rlinehan replied:
Yeah, that makes sense. I guess I hope someone would catch this in a review, but there's not a guarantee of that, and it's probably better to guard against it since it isn't hard.

camlow325 commented on src/puppetlabs/trapperkeeper/internal.clj:
```diff
@@ -210,15 +224,22 @@
+                      #{:boot :restart}
+                      (do
+                        (log/debugf "Lifecycle worker executing %s lifecycle task." type)
+                        (task-function)
```
It looks like the schema would allow for the `task-function` to be optional.  If that's intentional, should we check that `task-function` has a non nil value before executing it?  Assuming `task-function` should be required for a `:boot` or `:restart` task, should we harden the schema to ensure that it has a non nil value?

cprice404 replied:
Good catch.  In this PR, the 'shutdown' path doesn't have a task-function, which is why I had to do the schema that way.  In a later PR when I move the shutdown work into the go loop, the shutdown case gains a task-function.  So I think what would make sense to me is to leave this as-is in this PR but then just help make sure that I harden the schema in the follow-up PR?

camlow325 replied:
Fair enough, will try to remember to look for that.

camlow325 commented:
:+1: although it looks like we'll probably want to rebase this on top of the previously merged #222 since some of the changed code between this one and that one look pretty similar.

cprice404 commented:
Rebased and added a commit addressing @rlinehan 's comment from my previous PR.  I'm not sure I think there's anything immediately actionable from the other comments on this PR, but happy to revisit if anyone disagrees with my thoughts above.

rlinehan commented:
:+1: but looks like travis failed with some legitimate failures because of the log statement changing

```
FAIL in (test-max-queued-restarts) (internal_test.clj:89)

Missing expected log message when too many HUP requests queued

expected: (logged? (format "Too many SIGHUP restart requests queued (%d); ignoring!" internal/max-pending-lifecycle-events) :warn)

  actual: (logged [#<LoggingEvent [DEBUG] SIGHUP handler restarting TK apps.> #<LoggingEvent [WARN] Ignoring new SIGHUP restart requests; too many requests queued (5)>])
```

cprice404 commented:
dang it

cprice404 commented:
squashed in a fix for the test relating to the changed log message.

camlow325 commented:
Looks like Travis is passing and I don't see any other outstanding comments so merging this...
